gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java

```diff
@@ -0,0 +1,296 @@
+        for (Schema.Field field : f.schema().getFields()) {
+          flattenedFields.addAll(flattenField(field, lineage, level+1, false));
+        }
+        lineage.removeLast();
+      }
+    }
+    // If field.Type = OPTION, un-nest its fields and return them
+    else {
+      if (isOfOptionType(f.schema())) {
+        Schema first = f.schema().getTypes().get(0);
+        Schema second = f.schema().getTypes().get(1);
+        Schema record = first;
+        boolean wrapInOption = Schema.Type.NULL.equals(first.getType());
+        if (wrapInOption) {
+          record = second;
+        }
+        // Set lineage for wrapped Union
+        if (record.getFields().size() > 0) {
+          lineage.add(record.getName());
+          for (Schema.Field field : record.getFields()) {
+            flattenedFields.addAll(flattenField(field, lineage, level + 1, wrapInOption));
+          }
+          lineage.removeLast();
+        }
+      }
+      // If field.Type = any-other, copy and return it
+      else {
+        // Compute name and source using lineage
+        String flattenName = f.name();
+        String flattenSource = StringUtils.EMPTY;
+        if (lineage.size() > level + 1) {
+          // If level and lineage are not in sync, we have un-nested something
+          flattenName = StringUtils.join(lineage, "_");
+          flattenSource = StringUtils.join(lineage, ".");
+        }
+        // Copy field
+        Schema flattenedFieldSchema = flatten(f.schema(), level + 1);
+        if (shouldWrapInOption) {
+          // If already Union, just copy it instead of wrapping (Union within Union is not supported)
+          if (Schema.Type.UNION.equals(flattenedFieldSchema.getType())) {
+            List<Schema> newUnionMembers = new ArrayList<>();
+            newUnionMembers.add(Schema.create(Schema.Type.NULL)); // Its an Option, so first member must be NULL
+            for (Schema type : flattenedFieldSchema.getTypes()) {
+              if (Schema.Type.NULL.equals(type.getType())) {
+                continue; // Already added
+              }
+              newUnionMembers.add(type);
+            }
+            flattenedFieldSchema = Schema.createUnion(newUnionMembers);
+          }
+          // Wrap the Union, since parent Union is an option
+          else {
+            flattenedFieldSchema =
+                Schema.createUnion(Arrays.asList(Schema.create(Schema.Type.NULL), flattenedFieldSchema));
+          }
+        }
+        Schema.Field field = new Schema.Field(flattenName, flattenedFieldSchema, f.doc(), f.defaultValue(), f.order());
+
+        // Setup metadata
+        if (StringUtils.isNotBlank(flattenSource)) {
+          field.addProp("flatten_source", flattenSource);
+        }
+        for (Map.Entry<String, JsonNode> entry : f.getJsonProps().entrySet()) {
+          field.addProp(entry.getKey(), entry.getValue());
+        }
+        flattenedFields.add(field);
+      }
+    }
+    lineage.removeLast();
```

Comment:
Modifying "lineage" and trying to revert it to its original value is very error-prone. Someone will later modify the code, forget to do it and all kinds of bad things may happen. I think it is best to make "lineage" immutable and make copies if necessary to modify it.